src/tls/main.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <argp.h>
+#include <err.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <signal.h>
+
+#include <common/cockpitwebcertificate.h>
+#include "utils.h"
+#include "server.h"
+
+#define COCKPIT_WS PACKAGE_LIBEXEC_DIR "/cockpit-ws"
+
+/* CLI arguments */
+struct arguments {
+  uint16_t port;
+  int no_tls;
+  unsigned idle_timeout;
+};
+
+#define OPT_NO_TLS 1000
+#define OPT_IDLE_TIMEOUT 1001
+
+static unsigned
+arg_parse_unsigned (char *arg, struct argp_state *state, unsigned min, unsigned max, const char *error_msg)
+{
+  char *endptr = NULL;
+  long num = strtol (arg, &endptr, 10);
+
+  if (!*arg || *endptr != '\0' || num < min || num > max)
+    argp_error (state, "%s: %s", error_msg, arg);
+  return (unsigned) num;
+}
+
+/* Parse a single option. */
+static error_t
+parse_opt (int key, char *arg, struct argp_state *state)
+{
+  struct arguments *arguments = state->input;
+
+  switch (key)
+    {
+      case OPT_NO_TLS:
+        arguments->no_tls = 1;
+        break;
+      case 'p':
+        arguments->port = arg_parse_unsigned (arg, state, 1, UINT16_MAX, "Invalid port");
+        break;
+      case OPT_IDLE_TIMEOUT:
+        arguments->idle_timeout = arg_parse_unsigned (arg, state, 0, (UINT_MAX / 1000) - 1, "Invalid idle timeout") * 1000;
+        break;
+      default:
+        return ARGP_ERR_UNKNOWN;
+    }
+  return 0;
+}
+
+static struct argp_option options[] = {
+  {"no-tls", OPT_NO_TLS, 0, 0,  "Don't use TLS" },
+  {"port", 'p', "PORT", 0, "Local port to bind to (9090 if unset)" },
+  {"idle-timeout", OPT_IDLE_TIMEOUT, "SECONDS", 0, "Time after which to exit if there are no connections; 0 to run forever (default: 90)" },
+  { 0 }
+};
+
+static const struct argp argp = {
+  .options = options,
+  .parser = parse_opt,
+  .doc = "cockpit-tls -- TLS terminating proxy for cockpit-ws",
+};
+
+Server *server;  /* global, can't pass it to our signal handler */
+
+static void
+handle_sigchld (int signal, siginfo_t *info, void *ucontext)
```

Comment:
This is where life starts to suck.  Non-realtime signals can be coalesced by the kernel if multiple instances of them are outstanding.  It's therefore unsafe to assume that you will see the pids of all exiting children showing up in `si_pid`.

Best bet here is to ignore the pid and call non-blocking `waitpid(-1)` in a loop on each SIGCHLD.